### PR TITLE
[4.0] Add basic contact api

### PR DESCRIPTION
**Pull request by wilsonge**
Add a basic API plugin and related classes for com_contact. This is an example of what integration with the API is going to be like for 3rd party components too

### CODE REVIEW

File: administrator/components/com_admin/sql/updates/postgresql/4.0.0-2019-01-05.sql [hunk 1]
```diff
@@ -1,3 +1,4 @@
 INSERT INTO "#__extensions" ("extension_id", "name", "type", "element", "folder", "client_id", "enabled", "access", "protected", "manifest_cache", "params", "custom_data", "system_data", "checked_out", "checked_out_time", "ordering", "state") VALUES
 (494, 0, 'plg_api-authentication_basic', 'plugin', 'basic', 'api-authentication', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
 (495, 0, 'plg_webservices_content', 'plugin', 'content', 'webservices', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
+(496, 0, 'plg_webservices_contact', 'plugin', 'contact', 'webservices', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
```
Comment (brianteeman) [inline administrator/components/com_admin/sql/updates/postgresql/4.0.0-2019-01-05.sql]: 1970 on pgsql

File: administrator/language/en-GB/en-GB.plg_webservices_contact.ini [hunk 2]
```diff
@@ -0,0 +1,7 @@
+; Joomla! Project
+; Copyright (C) 2005 - 2017 Open Source Matters. All rights reserved.
+; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
+; Note : All ini files need to be saved as UTF-8
+
+PLG_WEBSERVICES_CATEGORIES="Web Services - Contacts"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_webservices_contact.ini]: CATEGORIES ???

File: administrator/language/en-GB/en-GB.plg_webservices_contact.sys.ini [hunk 3]
```diff
@@ -0,0 +1,7 @@
+; Joomla! Project
+; Copyright (C) 2005 - 2017 Open Source Matters. All rights reserved.
+; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
+; Note : All ini files need to be saved as UTF-8
+
+PLG_WEBSERVICES_CATEGORIES="Web Services - Contacts"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_webservices_contact.sys.ini]: CATEGORIES??

File: api/components/com_contact/Controller/ContactsController.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
+<?php
+/**
+ * @package     Joomla.API
+ * @subpackage  com_content
```
Comment (brianteeman) [inline api/components/com_contact/Controller/ContactsController.php]: shouldnt that be com_contact

File: api/components/com_contact/View/Contacts/JsonapiView.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
+<?php
+/**
+ * @package     Joomla.API
+ * @subpackage  com_content
```
Comment (brianteeman) [inline api/components/com_contact/View/Contacts/JsonapiView.php]: contact?

File: api/components/com_contact/Controller/ContactsController.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+namespace Joomla\Component\Contact\Api\Controller;
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\MVC\Controller\ApiController;
+
+/**
+ * The category controller
```
Comment (brianteeman) [inline api/components/com_contact/Controller/ContactsController.php]: contacts

File: api/components/com_contact/View/Contacts/JsonapiView.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+namespace Joomla\Component\Contact\Api\View\Contacts;
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\MVC\View\JsonApiView as BaseApiView;
+
+/**
+ * The article controller
```
Comment (brianteeman) [inline api/components/com_contact/View/Contacts/JsonapiView.php]: contacts

File: installation/sql/postgresql/joomla.sql [hunk 8]
```diff
@@ -687,6 +687,7 @@ INSERT INTO "#__extensions" ("extension_id", "package_id", "name", "type", "elem
 (493, 0, 'plg_extension_finder', 'plugin', 'finder', 'extension', 0, 1, 1, 0, '', '', 0, '1970-01-01 00:00:00', 0, 0),
 (494, 0, 'plg_api-authentication_basic', 'plugin', 'basic', 'api-authentication', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
 (495, 0, 'plg_webservices_content', 'plugin', 'content', 'webservices', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
+(496, 0, 'plg_webservices_contact', 'plugin', 'contact', 'webservices', 0, 1, 1, 0, '', '{}', 0, '0000-00-00 00:00:00', 0, 0),
```
Comment (brianteeman) [inline installation/sql/postgresql/joomla.sql]: 1970

File: plugins/webservices/contact/contact.php [hunk 9]
```diff
@@ -0,0 +1,43 @@
+<?php
+/**
+ * @package     Joomla.Plugin
+ * @subpackage  Webservices.Content
```
Comment (brianteeman) [inline plugins/webservices/contact/contact.php]: contact

File: plugins/webservices/contact/contact.php [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\Plugin\CMSPlugin;
+use Joomla\CMS\Router\ApiRouter;
+
+/**
+ * Web Services adapter for com_categories.
```
Comment (brianteeman) [inline plugins/webservices/contact/contact.php]: com_categorues?

File: plugins/webservices/contact/contact.php [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+	/**
+	 * Load the language file on instantiation.
+	 *
+	 * @var    boolean
+	 * @since  __DEPLOY_VERSION__
+	 */
+	protected $autoloadLanguage = true;
+
+	/**
+	 * Registers com_categories's API's routes in the application
```
Comment (brianteeman) [inline plugins/webservices/contact/contact.php]: com_categories?

File: plugins/webservices/contact/contact.xml [hunk 12]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
+<extension version="4.0" type="plugin" group="webservices" method="upgrade">
+	<name>plg_webservices_contact</name>
+	<author>Joomla! Project</author>
+	<creationDate>August 2017</creationDate>
```
Comment (brianteeman) [inline plugins/webservices/contact/contact.xml]: are we going back in time

**Review by brianteeman — CHANGES_REQUESTED** [review]
just reviewing mainly copy paste errors

less haste more speed ;)

**Comment by wilsonge** [comment]
Whoops - thanks. I initially wrote a categories implementation then changed my mind as it’s more complicated (should it be child endpoints of categories etc.). So swapped to contacts. Apparently badly :)

**Comment by brianteeman** [comment]
Confirming that the changes I requested have all been made.

However I dont have the skillset to test the actual code :(

**Comment by astridx** [comment]
Could it be that we have to change lines 23 and 24 in the file 
`libraries/src/Router/ApiRouter.php `
to 
```
use Joomla\CMS\Router\Router;
use Joomla\CMS\Router\Route;
```
(we have to add the `CMS`)
https://github.com/joomla/joomla-cms/blob/3084b818100d3972f76eba6d2d68fb99f890c075/libraries/src/Router/ApiRouter.php#L15

**Comment by mbabker** [comment]
@astridx see https://github.com/joomla/joomla-cms/pull/24135 - TL;DR, it's right as is

**Comment by alikon** [comment]
- `curl -X GET -H 'Authorization: Basic YWRtaW46YWRtaW4=' -i 'http://localhost/quattro/api/index.php/v1/contact'`
-  ok
- `curl -X GET -H 'Authorization: Basic YWRtaW46YWRtaW4=' -i 'http://localhost/quattro/api/index.php/v1/contact/1'`
- ok `{"links":{"self":"http:\/\/localhost\/quattro\/api\/index.php\/v1\/contact\/1"},"data":{"type":"contacts","id":"1","attributes":{"id":1,"name":"Contact","published":1,"user_id":0,"catid":4,"language":"*"}}}`

- `curl -X DELETE -H 'Authorization: Basic YWRtaW46YWRtaW4=' -i 'http://localhost/quattro/api/index.php/v1/contacts/1'`
-  KO  `{"errors":[{"code":500,"title":"Internal server error"}]}`

**Comment by wilsonge** [comment]
Closing as this has been integrated into the GSOC 19 webservices project

**Pull request closed without merging**